modules/tensor_mechanics/src/materials/IsotropicPowerLawHardeningStressUpdate.C

```diff
@@ -39,62 +41,73 @@ IsotropicPowerLawHardeningStressUpdate::validParams()
+  this->_plastic_strain[_qp] = this->_plastic_strain_old[_qp];
 }
 
-Real
-IsotropicPowerLawHardeningStressUpdate::computeHardeningDerivative(Real scalar)
+template <bool is_ad>
+GenericReal<is_ad>
+IsotropicPowerLawHardeningStressUpdateTempl<is_ad>::computeHardeningDerivative(
+    const GenericReal<is_ad> & scalar)
 {
-  const Real stress_delta = _effective_trial_stress - _three_shear_modulus * scalar;
-  Real slope = std::pow(stress_delta, (1.0 / _strain_hardening_exponent - 1.0)) /
-               _strain_hardening_exponent * 1.0 / std::pow(_K, 1.0 / _strain_hardening_exponent);
+  const GenericReal<is_ad> stress_delta = _effective_trial_stress - _three_shear_modulus * scalar;
+  GenericReal<is_ad> slope = std::pow(stress_delta, (1.0 / _strain_hardening_exponent - 1.0)) /
+                             _strain_hardening_exponent * 1.0 /
+                             std::pow(_K, 1.0 / _strain_hardening_exponent);
   slope -= 1.0 / _youngs_modulus;
 
   return 1.0 / slope;
 }
 
+template <bool is_ad>
 void
-IsotropicPowerLawHardeningStressUpdate::computeYieldStress(const RankFourTensor & elasticity_tensor)
+IsotropicPowerLawHardeningStressUpdateTempl<is_ad>::computeYieldStress(
+    const GenericRankFourTensor<is_ad> & elasticity_tensor)
 {
   // Pull in the Lam\`{e} lambda, and caculate E
-  const Real lambda = getIsotropicLameLambda(elasticity_tensor);
-  const Real shear_modulus = _three_shear_modulus / 3.0;
+  const GenericReal<is_ad> lambda = getIsotropicLameLambda(elasticity_tensor);
+  const GenericReal<is_ad> shear_modulus = _three_shear_modulus / 3.0;
 
   _youngs_modulus = shear_modulus * (3.0 * lambda + 2 * shear_modulus) / (lambda + shear_modulus);
 
   // Then solve for yield stress using equation from the header file
-  _yield_stress = std::pow(_K / std::pow(_youngs_modulus, _strain_hardening_exponent),
-                           1.0 / (1.0 - _strain_hardening_exponent));
-  if (_yield_stress <= 0.0)
+  this->_yield_stress = std::pow(_K / std::pow(_youngs_modulus, _strain_hardening_exponent),
+                                 1.0 / (1.0 - _strain_hardening_exponent));
+  if (this->_yield_stress <= 0.0)
     mooseError("The yield stress must be greater than zero, but during the simulation your yield "
                "stress became less than zero.");
 }
 
-Real
-IsotropicPowerLawHardeningStressUpdate::getIsotropicLameLambda(
-    const RankFourTensor & elasticity_tensor)
+template <bool is_ad>
+GenericReal<is_ad>
+IsotropicPowerLawHardeningStressUpdateTempl<is_ad>::getIsotropicLameLambda(
+    const GenericRankFourTensor<is_ad> & elasticity_tensor)
 {
-  const Real lame_lambda = elasticity_tensor(0, 0, 1, 1);
+  const GenericReal<is_ad> lame_lambda = elasticity_tensor(0, 0, 1, 1);
 
-  if (_mesh.dimension() == 3 && lame_lambda != elasticity_tensor(1, 1, 2, 2))
+  if (this->_mesh.dimension() == 3 &&
+      MetaPhysicL::raw_value(lame_lambda) != MetaPhysicL::raw_value(elasticity_tensor(1, 1, 2, 2)))
```

Comment:
This should be using the Guarantee system. But that can be a follow on PR

Reply:
I just tried to do this and I can't.  I would need the _elasticity_tensor_name and I don't derive off of any classes with that defined.

Reply:
ok, got it! It would have to be in the calling class. Something we can address later